extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/PanacheQuery.java

```diff
@@ -40,6 +41,19 @@
      */
     public <T> PanacheQuery<T> project(Class<T> type);
 
+    /**
+     * Generates a SELECT clause with the specified fields.
+     * <p>
+     * The result mapper function is used to build a DTO object from a result object. A result
+     * object is represented either by {@code Object[]} for multiple fields or {@code Object} for a single field.
+     * 
+     * @param <T>
+     * @param fields
+     * @param resultMapper
+     * @return a new query with the same state as the previous one (params, page, range, lockMode, hints, ...).
+     */
+    public <T> PanacheQuery<T> select(List<String> fields, Function<Object, T> resultMapper);
```

Comment:
If we change the order of the parameters, we can get a more type-safe API:

```java
public <T> PanacheQuery<T> select(Function<Object, T> rowMapper, String field);
public <T> PanacheQuery<T> select(Function<Object[], T> rowMapper, String... fields);
```

It's a bit backwards, but it saves having to cast to an array when using more than one field. WDYT?

Reply:
I've decided to only add  `public <T> PanacheQuery<T> project(Function<Object[], T> rowMapper, String... properties)` so that a result row is always represented by `Object[]`. I think that it's a more clear and "unified" approach.